File: src/telemetry.rs
use std::env;
use std::fs;
use std::path::PathBuf;

use posthog_rs::{ClientOptionsBuilder, Event};

const POSTHOG_API_KEY: &str = "phc_RpuEAMMomACJxc7hG4mRMKURklt2BXtfzzwYQYlzr0W";

pub struct Telemetry {
    client: posthog_rs::Client,
    distinct_id: Option<String>,
}

impl Telemetry {
    pub async fn new() -> Self {
        let disabled = cfg!(debug_assertions)
            || matches!(
                env::var("ORIYN_TELEMETRY").as_deref(),
                Ok("0" | "false" | "off")
            )
            || config_dir().join("telemetry-disabled").exists();

        let options = ClientOptionsBuilder::default()
            .api_key(POSTHOG_API_KEY.to_string())
            .disabled(disabled)
            .build()
            .expect("valid posthog options");

        let client = posthog_rs::client(options).await;
        let distinct_id = load_user_id().or_else(load_or_create_anonymous_id);

        Self { client, distinct_id }
    }

    pub async fn capture(&self, event_name: &str, props: serde_json::Value) {
        let Some(ref did) = self.distinct_id else {
            return;
        };
        let mut event = Event::new(event_name, did);
        event.insert_prop("$lib", "oriyn-cli").ok();
        event
            .insert_prop("cli_version", env!("CARGO_PKG_VERSION"))
            .ok();
        event.insert_prop("$os", env::consts::OS).ok();
        if let Some(obj) = props.as_object() {
            for (k, v) in obj {
                event.insert_prop(k, v).ok();
            }
        }
        let _ = self.client.capture(event).await;
    }
}

fn config_dir() -> PathBuf {
    let home = env::var("HOME").unwrap_or_else(|_| ".".to_string());
    PathBuf::from(home).join(".config").join("oriyn")
}

pub fn get_user_id() -> Option<String> {
    load_user_id()
}

fn load_user_id() -> Option<String> {
    fs::read_to_string(config_dir().join("user-id"))
        .ok()
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty())
}

fn load_or_create_anonymous_id() -> Option<String> {
    let path = config_dir().join("anonymous-id");
    if let Ok(id) = fs::read_to_string(&path) {
        let id = id.trim().to_string();
        if !id.is_empty() {
            return Some(id);
        }
    }
    let id = uuid::Uuid::new_v4().to_string();
    fs::create_dir_all(config_dir()).ok()?;
    fs::write(&path, &id).ok()?;
    Some(id)
}

pub fn store_user_id(user_id: &str) {
    let dir = config_dir();
    let _ = fs::create_dir_all(&dir);
    let _ = fs::write(dir.join("user-id"), user_id);
}

pub fn clear_user_id() {
    let _ = fs::remove_file(config_dir().join("user-id"));
}

/// Manage telemetry opt-in/out.
pub fn manage(disable: bool, enable: bool, status: bool) {
    let flag_path = config_dir().join("telemetry-disabled");

    if disable {
        let _ = fs::create_dir_all(config_dir());
        let _ = fs::write(&flag_path, "");
        println!("Telemetry disabled.");
    } else if enable {
        let _ = fs::remove_file(&flag_path);
        println!("Telemetry enabled.");
    } else if status {
        let disabled = cfg!(debug_assertions)
            || matches!(
                env::var("ORIYN_TELEMETRY").as_deref(),
                Ok("0" | "false" | "off")
            )
            || flag_path.exists();
        if disabled {
            println!("Telemetry: disabled");
        } else {
            println!("Telemetry: enabled");
        }
    }
}

File: internal/telemetry/telemetry.go
package telemetry

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"

	"github.com/google/uuid"
	"github.com/posthog/posthog-go"
)

// PostHog write-only client key — safe to ship in OSS binaries; equivalent
// to a public publishable key. See posthog.com/docs/api#authentication.
//
//nolint:gosec // G101: not a credential — public PostHog client key.
const posthogAPIKey = "phc_RpuEAMMomACJxc7hG4mRMKURklt2BXtfzzwYQYlzr0W"

type Tracker struct {
	client     posthog.Client
	distinctID string
	version    string
}

func NewTracker(version string) *Tracker {
	disabled := version == "dev" || isEnvDisabled() || sentinelFileExists()

	t := &Tracker{version: version}

	if !disabled {
		client, err := posthog.NewWithConfig(posthogAPIKey, posthog.Config{})
		if err == nil {
			t.client = client
		}
	}

	if id := loadUserID(); id != "" {
		t.distinctID = id
	} else {
		t.distinctID = loadOrCreateAnonymousID()
	}

	return t
}

func (t *Tracker) Capture(event string, props map[string]interface{}) {
	if t.client == nil || t.distinctID == "" {
		return
	}
	p := posthog.NewProperties()
	p.Set("$lib", "oriyn-cli")
	p.Set("cli_version", t.version)
	p.Set("$os", runtime.GOOS)
	for k, v := range props {
		p.Set(k, v)
	}
	_ = t.client.Enqueue(posthog.Capture{
		DistinctId: t.distinctID,
		Event:      event,
		Properties: p,
	})
}

func (t *Tracker) Close() {
	if t.client != nil {
		_ = t.client.Close()
	}
}

func configDir() string {
	home, err := os.UserHomeDir()
	if err != nil {
		home = "."
	}
	return filepath.Join(home, ".config", "oriyn")
}

func StoreUserID(id string) {
	dir := configDir()
	_ = os.MkdirAll(dir, 0o700)
	_ = os.WriteFile(filepath.Join(dir, "user-id"), []byte(id), 0o600)
}

func ClearUserID() {
	_ = os.Remove(filepath.Join(configDir(), "user-id"))
}

func GetUserID() string {
	return loadUserID()
}

func loadUserID() string {
	data, err := os.ReadFile(filepath.Join(configDir(), "user-id"))
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(data))
}

func loadOrCreateAnonymousID() string {
	path := filepath.Join(configDir(), "anonymous-id")
	data, err := os.ReadFile(path)
	if err == nil {
		if id := strings.TrimSpace(string(data)); id != "" {
			return id
		}
	}
	id := uuid.NewString()
	_ = os.MkdirAll(configDir(), 0o700)
	_ = os.WriteFile(path, []byte(id), 0o600)
	return id
}

func isEnvDisabled() bool {
	v := strings.ToLower(os.Getenv("ORIYN_TELEMETRY"))
	return v == "0" || v == "false" || v == "off"
}

func sentinelFileExists() bool {
	_, err := os.Stat(filepath.Join(configDir(), "telemetry-disabled"))
	return err == nil
}

func Manage(disable, enable, status bool, version string) {
	flagPath := filepath.Join(configDir(), "telemetry-disabled")

	switch {
	case disable:
		_ = os.MkdirAll(configDir(), 0o700)
		_ = os.WriteFile(flagPath, []byte(""), 0o600)
		fmt.Println("Telemetry disabled.")
	case enable:
		_ = os.Remove(flagPath)
		fmt.Println("Telemetry enabled.")
	case status:
		disabled := version == "dev" || isEnvDisabled() || sentinelFileExists()
		if disabled {
			fmt.Println("Telemetry: disabled")
		} else {
			fmt.Println("Telemetry: enabled")
		}
	}
}
